### ISSUE TITLE
Customizable representation type

### ISSUE THREAD

**Issue by dscottboggs** [dscottboggs/rust-classic-bitfield#1]
The bitfield enum desugars into a singleton struct which contains a `u64`.

```rust
struct SomeBitfieldu64);
```

and the "variants" are constants defined under an impl:

```rust
impl SomeBitfield {
    const VARIANT: u64 = 0u64;
}
```

I'd like this underlying type, `.0`, to be able to be represented by other types, not just `u64`, like this:

```rust
#[bitfield_enum]
#[repr(as u8)]
enum SomeBitfield {
    VARIANT_ONE,
}
// *SomeBitfield::VARIANT_ONE should be 1u8
```

Really, for the library to be useful, I think it needs to be. But I ran into a roadblock. The `0u64` in the assignment is important, it's inserted implicitly by `quote!`:

```rust
...
self.variants
    .iter()
    .enumerate()
    .map(|(i, variant)|
        let i = 1u64 << i;
        ...
        let repr = ....
            .unwrap_or(quote! { #i });
...
```
[(see the full code here)](https://github.com/dscottboggs/rust-classic-bitfield/blob/abc3d5c618a392e5cb8ea1b3e55050fa58d3e52e/src/context.rs#L65-L93)

Here we've specified the value as `1u64` to give a sane default, but that would need replaced with whatever the type `self.repr_type` represents, even if that's not u64.

### Runtime conversion
We could define each constant as

```rust
static #name: LazyLock<#repr_type> = LazyLock::new(|| #repr_type::from(#repr))
```

However, this requires the use of the nightly compiler and comes with some runtime overhead which may be undesirable. It also precludes the possibility of a `no-std` feature.

There is also the `lazy` crate, but that would require a peer dependency, which cargo doesn't seem to support, and still incurs that runtime overhead.

### Another macro
We should also be able to pass `repr_type` and `i` into some procedural macro which should be able to spit back out the appropriate value, which will require another crate, and the approach here is not yet clear. Research is in progress on this. `macro_rules` macros don't seem to be an option here.

**Comment by dscottboggs** [comment]
I suspect that this will work as a workaround since the value is specified as the right type on each variant, but I haven't tested it yet:

```rust
#[bitfield_enum]
#[repr(as u8)]
enum Permissions {
    #[repr(1_u8)]
    EXECUTE,
    #[repr(2_u8)]
    WRITE,
    #[repr(4_u8)]
    READ,
}
```

This is definitely a workaround though, I don't like the ergonomics.

**Comment by dscottboggs** [comment]
resolved in [b69f14c037](https://github.com/dscottboggs/rust-classic-bitfield/commit/b69f14c037d8a04a40668361a2f25cb61d8c8d3e)
